### PR TITLE
normalize loaded build packages and handle missing packages without panicking

### PR DESCRIPTION
**Pull request by liggitt**
/hold for clean run in k/k

### CODE REVIEW

File: parser/parse.go [hunk 1]
```diff
@@ -118,11 +120,37 @@ func (b *Builder) AddBuildTags(tags ...string) {
 	b.context.BuildTags = append(b.context.BuildTags, tags...)
 }
 
+func (b *Builder) getLoadedBuildPackage(importPath string) (*build.Package, bool) {
+	canonicalized := canonicalizeImportPath(importPath)
+	if string(canonicalized) != string(importPath) {
```
Comment (thockin) [inline parser/parse.go:125]: why are you casting `importPath` to string ?  It's already a string...

File: parser/parse.go [hunk 2]
```diff
@@ -118,11 +120,37 @@ func (b *Builder) AddBuildTags(tags ...string) {
...
+		klog.V(5).Infof("getLoadedBuildPackage: %s normalized to %s", importPath, canonicalized)
+	}
+	buildPkg, ok := b.buildPackages[canonicalized]
+	return buildPkg, ok
+}
+func (b *Builder) setLoadedBuildPackage(importPath string, buildPkg *build.Package) {
+	canonicalizedImportPath := canonicalizeImportPath(importPath)
+	if string(canonicalizedImportPath) != string(importPath) {
```
Comment (thockin) [inline parser/parse.go:133]: Same - why cast?

File: parser/parse.go [hunk 3]
```diff
@@ -118,11 +120,37 @@ func (b *Builder) AddBuildTags(tags ...string) {
...
+		klog.V(5).Infof("setLoadedBuildPackage: importPath %s normalized to %s", importPath, canonicalizedImportPath)
+	}
+
+	canonicalizedBuildPkgImportPath := canonicalizeImportPath(buildPkg.ImportPath)
+	if string(canonicalizedBuildPkgImportPath) != string(buildPkg.ImportPath) {
```
Comment (thockin) [inline parser/parse.go:138]: This cast also feels suspect - are they both `importPathString`?  Why not just compare them?

File: parser/parse.go [hunk 4]
```diff
@@ -118,11 +120,37 @@ func (b *Builder) AddBuildTags(tags ...string) {
...
+		klog.V(5).Infof("setLoadedBuildPackage: buildPkg.ImportPath %s normalized to %s", buildPkg.ImportPath, canonicalizedBuildPkgImportPath)
+	}
+
+	if string(canonicalizedImportPath) != string(canonicalizedBuildPkgImportPath) {
```
Comment (thockin) [inline parser/parse.go:142]: same

**Review by thockin — COMMENTED** [review]
I think I grok what this is doing.  We have a distinct lack of testing here, don't we...

File: parser/parse.go (hunk 3 above)
Comment (liggitt) [reply parser/parse.go:138]: `buildPkg.ImportPath` is a string (it's not our type), but I dropped the unnecessary cast

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

File: parser/parse.go [hunk 5]
```diff
@@ -287,8 +313,8 @@ func (b *Builder) AddDirectoryTo(dir string, u *types.Universe) (*types.Package,
 	if _, err := b.importPackage(dir, true); err != nil {
 		return nil, err
 	}
-	pkg, ok := b.buildPackages[dir]
-	if !ok {
+	pkg, ok := b.getLoadedBuildPackage(dir)
+	if !ok || pkg == nil {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply parser/parse.go:317]: there were a bunch of places where we were looking up directly in the map and assuming non-nil... this was just one

**Pull request merged**